blogs/DistributedLock.md

```diff
@@ -0,0 +1,80 @@
+	<li>Managing a pool of shared out-of-process resources (e. g. processes or cache file space)</li>
+</ul>
+
+Recently, I faced just such a challenge which I solved by implementing a small <a href="https://www.nuget.org/packages/DistributedLock/">distributed locking library</a> for .NET. 
+
+<h3 id="problem">The Problem</h3>
+
+I was recently reworking some code that establishes a user's session in several legacy sub-systems of an application. Because this process can be expensive and isn't always needed, I wanted to do it lazily. The initialization code looked something like this:
+
+<pre>
+var existingSessionInfo = GetLegacySessionInformation(); // HTTP GET request to authentication service
+if (existingSessionInfo != null) {
+	return existingSessionInfo;
+}
+
+var token1 = LogIntoLegacySystem1(); // HTTP POST request to one legacy system
+
+var token2 = LogIntoLegacySystem2(token1); // HTTP POST request to another legacy system
+
+SaveLegacySessionInfo(new LegacySessionInfo(token1, token2)); // HTTP POST request to authentication service
+</pre>
+
+The problem, of course, is that this code isn't thread-safe. If two threads get past the initial check for information, both will end up trying to establish sessions for the same user. If all steps were thread-safe and idempotent we might still be okay. Unfortunately, as is so often true with legacy code, that was not the case here.
+
+<h3 id="the solution">The Solution</h3>
+
+It shouldn't come as a surprise that the solution I chose was to employ distributed locking. Using the <a href="https://www.nuget.org/packages/DistributedLock/">DistributedLock</a> library, here's what the code looks like:
+
+<pre>
+// don't bother taking the lock if we don't need to
+var existingSessionInfo = GetLegacySessionInformation(); // HTTP GET request to authentication service
+if (existingSessionInfo != null) {
+	return existingSessionInfo;
+}
+
+var distributedLock = new SqlDistributedLock(
+	// a distributed lock's name is it's identity. Our lock incorporates the user ID so
+	// that we are synchronizing access only for that one user
+	lockName: "LegacySessionInitializationFor" + userId, 
+	// ConnectionString here is a connection string to a SQLServer database. Note that 
+	// this need not be the same database as is used by the various services we're talking to!
+	connectionString: ConnectionString
+);
+
+using (distributedLock.Acquire())
+{
+	// check again in case someone did this while we were waiting for the lock
+	existingSessionInfo = GetLegacySessionInformation();
+	if (existingSessionInfo != null) {
+		return existingSessionInfo;
+	}
+
+	var token1 = LogIntoLegacySystem1(); // HTTP POST request to one legacy system
+
+	var token2 = LogIntoLegacySystem2(token1); // HTTP POST request to another legacy system
+
+	SaveLegacySessionInfo(new LegacySessionInfo(token1, token2)); // HTTP POST request to authentication service
+}
+</pre>
```

Comment:
I think this example is more complex than it needs to be.  Why not something like:

```
UserProfile userProfile;
userProfile = GetUserProfile(/*userId*/17);
if (userProfile != null) { return userProfile; }
var distributedLock = new SqlDistributedLock(
  // a distributed lock's name is it's identity. Our lock incorporates the user ID so
  // that we are synchronizing access only for that one user
  lockName: "LegacySessionInitializationFor" + userId, 
  // ConnectionString here is a connection string to a SQLServer database. Note that 
  // this need not be the same database as is used by the various services we're talking to!
  connectionString: ConnectionString
);
using (distributedLock.Acquire()) {
  // check again in case someone did this while we were waiting for the lock
  userProfile = GetUserProfile(/*userId*/17);
  if (userProfile != null) { return userProfile; }
  return CreateUserProfile(/*userId*/17);
}
```